File: debian/generate-pkgfiles.pl
#!/usr/bin/perl

use strict;
use warnings;
use feature ':5.10';
use autodie;
use FindBin '$Bin';
use IPC::Run 'run';


my $description_cpp = <<'EOF';
GNU C cross-preprocessor (cpp)
 This is the GNU C preprocessor cpp built for cross-building $DEB_TARGET_ARCH
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned cpp cross package and symlink to it.
EOF

my $description_gcc = <<'EOF';
GNU C cross-compiler
 This is the GNU C compiler built for cross-building $DEB_TARGET_ARCH
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.
EOF

my $description_gpp = <<'EOF';
GNU C++ cross-compiler
 This is the GNU C++ compiler built for cross-building $DEB_TARGET_ARCH
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.
EOF

my $description_gfortran = <<'EOF';
GNU Fortran 95 cross-compiler
 This is the GNU Fortran 95 compiler built for cross-building $DEB_TARGET_ARCH
 binaries/packages. This is actually a metapackage that will bring in
 the correct versioned gcc cross package and symlink to it.
EOF

my $description_gobjc = <<'EOF';
GNU objective C cross-compiler
 This is the GNU objective C compiler built for cross-building $DEB_TARGET_ARCH
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.
EOF

my $description_gccgo = <<'EOF';
GNU go cross-compiler
 This is the GNU go compiler built for cross-building $DEB_TARGET_ARCH
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.
EOF

my %base_descriptions = ( 'cpp'      => $description_cpp,
                          'gcc'      => $description_gcc,
                          'g++'      => $description_gpp,
                          'gfortran' => $description_gfortran,
                          'gobjc'    => $description_gobjc,
                          'gccgo'    => $description_gccgo );







my $target_list_str = $ENV{TARGET_LIST} || `cat $Bin/targetlist` || ' ';
my @target_list = split / /, $target_list_str or
  die "Couldn't get target list from the TARGET_LIST env var, or from the file '$Bin/targetlist'";

say "Generating debian/control for arches '@target_list'";

my @pkgs    = split(/ /, runchild(qw(make --quiet -f), "$Bin/rules", 'say_pkgs_release'));
my $release = pop @pkgs;

open my $fd_control_out, '>', "$Bin/control";

{
    open my $fd_control_head, '<', "$Bin/control.head.in";
    local $/ = undef;
    my $head = <$fd_control_head>;
    print $fd_control_out $head;
    close $fd_control_head;
}

for my $DEB_TARGET_ARCH (@target_list)
{
    my $DEB_TARGET_GNU_TYPE =
      runchild(qw(dpkg-architecture -qDEB_HOST_GNU_TYPE -f), "-a$DEB_TARGET_ARCH");
    $DEB_TARGET_GNU_TYPE =~ s/_/-/g;

    say $fd_control_out "";

    for my $pkg (@pkgs)
    {
        my $description;


        my $subst_into_control = sub
        {
            foreach(@_)
            {
                s/\$DEB_TARGET_GNU_TYPE/$DEB_TARGET_GNU_TYPE/;
                s/\$DEB_TARGET_ARCH/$DEB_TARGET_ARCH/;
                s/\$pkg/$pkg/;
                s/\$ver/$release/;
                s/\$description/$description/;
            }
        };



        $description = description($pkg, $DEB_TARGET_ARCH);
        $subst_into_control->($description);

        open my $fd_control_in, '<', "$Bin/control.pkg.in";
        while(<$fd_control_in>)
        {
            $subst_into_control->($_);
            print $fd_control_out $_;
        }
    }

    generate_alternatives($DEB_TARGET_GNU_TYPE);
}




sub description
{
    my $pkg  = shift;
    my $arch = shift;

    my $base;
    if( $base_descriptions{$pkg} )
    {
        $base = $base_descriptions{$pkg};
    }
    else
    {
        $base_descriptions{$pkg} = $base =
          runchild(qw(dpkg-query -f), '${Description}', '-W', $pkg);
    }

    my $description = $base;
    $description =~ s/$/ for architecture $arch/m;
    return $description;
}

sub runchild
{
    my @args = @_;

    my ($in,$out,$err) = ('','','');
    run \@args, \$in, \$out, \$err
      or die "Error running '@args'. STDERR:\n$err";

    chomp $out;
    if( length($out) <= 0)
    {
        die "Error running '@args': output empty. STDERR:\n$err";
    }

    return $out;
}

sub generate_alternatives
{
    my ($DEB_TARGET_GNU_TYPE) = @_;

    for my $prog (qw(gcc g++ gfortran))
    {
        for my $inputfile (map {"$Bin/$prog.alternatives.$_.in"} qw(prerm postinst))
        {
            my $outputfile = $inputfile;
            $outputfile =~ s/(.*)\.alternatives(.*)\.in/$1-$DEB_TARGET_GNU_TYPE$2/;

            open my $fd_in,  '<', $inputfile;
            open my $fd_out, '>', $outputfile;

            while (<$fd_in>)
            {
                s/\$DEB_TARGET_GNU_TYPE/$DEB_TARGET_GNU_TYPE/g;
                print $fd_out $_;
            }
        }
    }
}

File: debian/generate-pkgfiles.py
#!/usr/bin/env python3

from collections import OrderedDict
import sys, subprocess, re, os

PKG_DESC = {
"cpp": OrderedDict([
    ("Package", "%(name)s-%(DEB_TARGET_GNU_TYPE)s"),
    ("Architecture", "all"),
    ("Depends", "%(name)s-%(major)s-%(DEB_TARGET_GNU_TYPE)s (>= %(Version)s), ${misc:Depends}"),
    ("Description", """GNU C cross-preprocessor (cpp)
 This is the GNU C preprocessor cpp built for cross-building %(DEB_TARGET_ARCH)s
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned cpp cross package and symlink to it.""")
    ]),

"gcc": ([
    ("Package", "%(name)s-%(DEB_TARGET_GNU_TYPE)s"),
    ("Architecture", "all"),
    ("Provides", "c-compiler:%(DEB_TARGET_ARCH)s"),
    ("Depends", "%(name)s-%(major)s-%(DEB_TARGET_GNU_TYPE)s (>= %(Version)s), ${misc:Depends}"),
    ("Description", """GNU C cross-compiler
 This is the GNU C compiler built for cross-building %(DEB_TARGET_ARCH)s
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.""")
]),

"g++": ([
    ("Package", "%(name)s-%(DEB_TARGET_GNU_TYPE)s"),
    ("Architecture", "all"),
    ("Provides", "g++:%(DEB_TARGET_ARCH)s (= 4:%(Version)s), c++:%(DEB_TARGET_ARCH)s (= 4:%(Version)s), c++-compiler:%(DEB_TARGET_ARCH)s, c++abi2-dev:%(DEB_TARGET_ARCH)s, build-essential:%(DEB_TARGET_ARCH)s (= 12.1)"),
    ("Depends", "%(name)s-%(major)s-%(DEB_TARGET_GNU_TYPE)s (>= %(Version)s), ${misc:Depends}"),
    ("Description", """GNU C++ cross-compiler
 This is the GNU C++ compiler built for cross-building %(DEB_TARGET_ARCH)s
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.""")
    ]),

"gfortran": ([
    ("Package", "%(name)s-%(DEB_TARGET_GNU_TYPE)s"),
    ("Architecture", "all"),
    ("Provides", "fortran-compiler:%(DEB_TARGET_ARCH)s, fortran95-compiler:%(DEB_TARGET_ARCH)s, fortran77-compiler:%(DEB_TARGET_ARCH)s, gfortran-mod-14:%(DEB_TARGET_ARCH)s"),
    ("Depends", "%(name)s-%(major)s-%(DEB_TARGET_GNU_TYPE)s (>= %(Version)s), ${misc:Depends}"),
    ("Description", """GNU Fortran 95 cross-compiler
 This is the GNU Fortran 95 compiler built for cross-building %(DEB_TARGET_ARCH)s
 binaries/packages. This is actually a metapackage that will bring in
 the correct versioned gcc cross package and symlink to it.""")
    ]),

"gobjc": ([
    ("Package", "%(name)s-%(DEB_TARGET_GNU_TYPE)s"),
    ("Architecture", "all"),
    ("Provides", "objc-compiler:%(DEB_TARGET_ARCH)s"),
    ("Depends", "%(name)s-%(major)s-%(DEB_TARGET_GNU_TYPE)s (>= %(Version)s), ${misc:Depends}"),
    ("Description", """GNU objective C cross-compiler
 This is the GNU objective C compiler built for cross-building %(DEB_TARGET_ARCH)s
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.""")
    ]),

"gccgo": ([ 
    ("Package", "%(name)s-%(DEB_TARGET_GNU_TYPE)s"),
    ("Architecture", "all"),
    ("Provides", "go-compiler:%(DEB_TARGET_ARCH)s"),
    ("Depends", "%(name)s-%(major)s-%(DEB_TARGET_GNU_TYPE)s (>= %(Version)s), ${misc:Depends}"),
    ("Description", """GNU go cross-compiler
 This is the GNU go compiler built for cross-building %(DEB_TARGET_ARCH)s
 binaries/packages. This is actually a metapackage that will bring in the
 correct versioned gcc cross package and symlink to it.""")
    ])
}


class Field:
    def __init__(self, pkg, k, v):
        self.pkg = pkg
        self.name = k
        self.value = v

    def __repr__(self):
        return "%s: %s" % (self.name, self.pkg.expand(self.value))


reSUBST = re.compile(r".*%\(\w+\)s.*")

class Package:
    vals = {
    }
    def __init__(self):
        self._fields = {}
        self._vals = dict(self.vals)

    def __repr__(self):
        return "\n".join(ii.__repr__() for ii in self.fields)

    @property
    def fields(self):
        return [Field(self, k, v) for k, v in self._fields.items()]

    def expand(self, val):
        exp = val % self._vals
        if reSUBST.match(exp):
            return self.expand(exp)
        else:
            return exp


class DebianPackage(Package):
    vals = {
    }
    def __init__(self, name, desc, target=None, ver="5.2"):
        super(DebianPackage, self).__init__()
        _vals = self._vals
        _vals["major"], _vals["minor"] = ver.split(".", 1)
        _vals["name"] = name
        _vals["Version"] = "%(major)s.%(minor)s"

        # Visible fields
        self._fields = _fields = OrderedDict(desc)
        assert("Description" in self._fields)
        assert("Architecture" in self._fields)
        assert("Description" in self._fields)

        _fields.move_to_end("Description")

        # All fields (including hidden)
        _vals.update(_fields)
        _vals.update(self.get_deb_archs(target=target))

    def get_deb_archs(self, host=None, target=None):
        args = ["dpkg-architecture"]
        if host:
            args.append("-a%s" % host)
        if target:
            args.append("-A%s" % target)
        c = subprocess.run(args, stdout=subprocess.PIPE, check=True)
        lines = (l.split("=") for l in c.stdout.decode("ascii").split("\n") if l)
        return {k: v.replace("_", "-") for k, v in lines}
        
def get_env_default(k):
    def_envs = {
        "TARGET_LIST":      "amd64 i386",
    }
    if os.environ.get(k):
        v = os.environ[k]
    else:
        v = def_envs.get(k, "")
        sys.stderr.write("%s variable not set, using default: %s\n" % (k, v))
    
    return v 

def main():
    sys.stdout.write(open(os.path.join(os.path.dirname(sys.argv[0]), "control.head.in"), "r").read())
    print("\n")
    targets = get_env_default("TARGET_LIST").split()
    packages = get_env_default("PACKAGE_LIST").split()
    ver = get_env_default("VER")
    
    for target in targets:
        for pkg in packages:
            print("%s\n" % DebianPackage(pkg, PKG_DESC[pkg], target, ver))


if __name__ == "__main__":
    main()

